src/snowflake/cli/api/entities/application_package_entity.py

```diff
@@ -161,6 +163,36 @@ def action_drop(
             force_drop=force_drop,
         )
 
+    def action_validate(self, ctx: ActionContext):
+        model = self._entity_model
+        package_name = model.fqn.identifier
+        stage_fqn = f"{package_name}.{model.stage}"
+        if model.meta and model.meta.role:
+            package_role = model.meta.role
+        else:
+            package_role = ctx.default_role
+
+        def deploy_to_scratch_stage_fn():
+            self.action_deploy(
+                ctx=ctx,
+                prune=True,
+                recursive=True,
+                paths=[],
+                validate=False,
+                stage_fqn=model.scratch_stage,
+            )
+
+        self.validate_setup_script(
+            console=ctx.console,
+            package_name=package_name,
+            package_role=package_role,
+            stage_fqn=stage_fqn,
+            use_scratch_stage=True,
+            scratch_stage_fqn=model.scratch_stage,
+            deploy_to_scratch_stage_fn=deploy_to_scratch_stage_fn,
```

Comment:
what's the reason we need to pass in a function to deploy to the stage instead of validate deploying directly?

Reply:
I'll refactor and simply deploy here directly. A few PRs ago it wasn't easy to do, but now I can refactor `deploy()` to a stateless function and reuse it in both `snow app` and `snow ws` flows.

Reply:
OK I gave it a try, but the change is still not as trivial as I hoped it to be. I'll follow up with this refactor in a dedicated PR.
The main reason I pass this as a function is that the `validate()` needs to be able to deploy, which requires passing many additional values through multiple call levels. It also requires an additional refactor to make the bundle/deploy methods static, because the validate one is static too.